pkg/detectors/postgres/postgres.go

```diff
@@ -69,186 +139,87 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) ([]dete
-				verifiedHosts = append(verifiedHosts, host)
-			}
-		}
-		if len(verifiedHosts) > 0 {
-			return verifiedHosts
-		}
+func getDeadlineInSeconds(ctx context.Context) int {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		// Context does not have a deadline
+		return 0
 	}
 
-	return hosts
+	duration := time.Until(deadline)
+	return int(duration.Seconds())
 }
 
-// deduplicate matches in order to reduce the number of verification requests
-func dedupMatches(matches [][]string) []string {
-	setOfMatches := make(map[string]struct{})
-	for _, match := range matches {
-		if len(match) > 1 {
-			setOfMatches[match[1]] = struct{}{}
-		}
-	}
-	var results []string
-	for match := range setOfMatches {
-		results = append(results, match)
+func isErrorDatabaseNotFound(err error, dbName string) bool {
+	if dbName == "" {
+		dbName = "postgres"
 	}
-	return results
+	missingDbErrorText := fmt.Sprintf("database \"%s\" does not exist", dbName)
+
+	return strings.Contains(err.Error(), missingDbErrorText)
 }
 
-func verifyPostgres(pgURL *url.URL, timeoutInSeconds int) (bool, error) {
-	if pgURL.User == nil {
-		return false, nil
-	}
-	username := pgURL.User.Username()
-	password, _ := pgURL.User.Password()
+func verifyPostgres(params map[string]string) (bool, error) {
+	if sslmode := params[pg_sslmode]; sslmode == pg_sslmode_allow || sslmode == pg_sslmode_prefer {
+		// pq doesn't support 'allow' or 'prefer'. If we find either of them, we'll just ignore it. This will trigger
+		// the same logic that is run if no sslmode is set at all (which mimics 'prefer', which is the default).
+		delete(params, pg_sslmode)
 
-	hostname, port := pgURL.Hostname(), pgURL.Port()
-	if hostname == "" {
-		hostname = defaultHost
-	}
-	if port == "" {
-		port = defaultPort
+		// We still want to save the original sslmode in ExtraData, so we'll re-add it before returning.
+		defer func() {
+			params[pg_sslmode] = sslmode
+		}()
 	}
 
-	sslmode := determineSSLMode(pgURL)
-
-	connStr := fmt.Sprintf("user=%s password=%s host=%s port=%s sslmode=%s", username, password, hostname, port, sslmode)
-	if timeoutInSeconds > 0 {
-		connStr = fmt.Sprintf("%s connect_timeout=%d", connStr, timeoutInSeconds)
+	var connStr string
+	for key, value := range params {
+		connStr += fmt.Sprintf("%s='%s'", key, value)
 	}
 
 	db, err := sql.Open("postgres", connStr)
 	if err != nil {
-		if strings.Contains(err.Error(), "connection refused") {
-			// inactive host
-			return false, nil
-		}
 		return false, err
 	}
 	defer db.Close()
 
-	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
-	defer cancel()
-
-	err = db.PingContext(ctx)
+	err = db.Ping()
 	if err == nil {
 		return true, nil
-	} else if strings.Contains(err.Error(), "password authentication failed") || // incorrect username or password
-		strings.Contains(err.Error(), "connection refused") { // inactive host
+	} else if strings.Contains(err.Error(), "password authentication failed") {
```

Comment:
nit: A switch is a lot easier to read here imo. We can also capture the final `return false, nil` in the default block:
```
switch {
	case err == nil:
		return true, nil
	case strings.Contains(err.Error(), "password authentication failed"):
		return false, nil
	case errors.Is(err, pq.ErrSSLNotSupported) && params[pg_sslmode] == "":
		// If the sslmode is unset, then either it was unset in the candidate secret, or we've intentionally unset it
		// because it was specified as 'allow' or 'prefer', neither of which pq supports. In all of these cases, non-SSL
		// connections are acceptable, so now we try a connection without SSL.
		params[pg_sslmode] = pg_sslmode_disable
		defer delete(params, pg_sslmode) // We want to return with the original params map intact (for ExtraData)
		return verifyPostgres(params)
	case isErrorDatabaseNotFound(err, params[pg_dbname]):
		return true, nil // If we know this, we were able to authenticate
	default:
		return false, nil
	}
```